google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java

```diff
@@ -125,20 +142,57 @@ void setPrevTransactionId(ByteString transactionId) {
       options = options.toBuilder().setReadWrite(readWrite).build();
     }
 
+    private io.opentelemetry.api.trace.Span startSpanWithParentContext(
+        String spanName,
+        com.google.cloud.datastore.telemetry.TraceUtil.SpanContext parentSpanContext) {
+      com.google.cloud.datastore.telemetry.TraceUtil otelTraceUtil =
+          datastore.getOptions().getTraceUtil();
+      SpanBuilder spanBuilder =
+          otelTraceUtil
+              .getTracer()
+              .spanBuilder(com.google.cloud.datastore.telemetry.TraceUtil.SPAN_NAME_TRANSACTION_RUN)
+              .setSpanKind(SpanKind.PRODUCER)
+              .setParent(
+                  Context.current()
+                      .with(
+                          io.opentelemetry.api.trace.Span.wrap(
+                              parentSpanContext.getSpanContext())));
+      return otelTraceUtil.addSettingsAttributesToCurrentSpan(spanBuilder).startSpan();
+    }
+
     @Override
     public T call() throws DatastoreException {
-      transaction = datastore.newTransaction(options);
-      try {
+      // TODO Instead of using OTel Spans directly, TraceUtil.Span should be used here. However,
+      // the same code in startSpanInternal doesn't work when EnabledTraceUtil.StartSpan is called
+      // probably because of some thread-local caching that is getting lost. This needs more
+      // debugging. The code below works and is idiomatic but could be prettier and more consistent
+      // with the use of TraceUtil-provided framework.
+      io.opentelemetry.api.trace.Span span =
```

Comment:
- Looks like ReadWriteTransactionCallable is scheduled to run on an Executor in the `RetryHelper` library provided by gax layer. So that complicates the Context propagation.
- Also tried "manually" propagating the parent SpanContext to the "callable" but that doesn't work either as we don't have control over the future lifetimes (which run in the RetryHelper) so the parent span may `end()` before the child spans are executed(?).
- This needs some significant overhauling or gax support for context propagation - both are non trivial and wouldn't want to block the launch on this.

Reply:
Uh... not sure why using the TraceUtil directly doesn't work. Maybe it's because there's only 1 TraceUtil but several threads creating spans? If that's the case, can you try and see if this works:

Add an API `TraceUtil.isTracingEnabled()`, and then change:

```java
private TraceUtil.Span startSpanWithParentContext(
        String spanName,
        TraceUtil.SpanContext parentSpanContext) {
  TraceUtil otelTraceUtil = datastore.getOptions().getTraceUtil();
  if (!otelTraceUtil.isTracingEnabled()) {
    return new TraceUtil.Span();
  }

  SpanBuilder spanBuilder =
      otelTraceUtil
          .getTracer()
          .spanBuilder(com.google.cloud.datastore.telemetry.TraceUtil.SPAN_NAME_TRANSACTION_RUN)
          .setSpanKind(SpanKind.PRODUCER)
          .setParent(
              Context.current()
                  .with(
                      io.opentelemetry.api.trace.Span.wrap(
                          parentSpanContext.getSpanContext())));
  return TraceUtil.Span(otelTraceUtil.addSettingsAttributesToCurrentSpan(spanBuilder).startSpan());
}
```

```java
TraceUtil.Span span =  startSpanWithParentContext(
              com.google.cloud.datastore.telemetry.TraceUtil.SPAN_NAME_TRANSACTION_RUN,
              parentSpanContext);

try (TraceUtil.Scope ignored = span.makeCurrent()) {
...
```

This way the opentelemetry code will remain off the execution path if tracing is disabled, and OTEL types are also not used on the main code path.